content/posts/2021/10/libplanet-0.11-0.19/index.eng.md

```diff
@@ -0,0 +1,251 @@
+the [`NetMQTransport`][NetMQTransport] which implements it is built-in.
+
+As it is a pilot phase, the interface still has been adjusted through several
+minor releases, and game programmers can't yet use their own transport
+implementation.  Later in the year, [`Swarm<T>`][Swarm<T>] instances will be
+configurable with custom transports besides
+the [new TCP-based transport][TcpTransport], which aims to replace
+the `NetMQTransport`.
+
+[ZeroMQ]: https://zeromq.org/
+[NetMQ]: https://github.com/zeromq/netmq
+[ITransport]: https://docs.libplanet.io/0.19.0/api/Libplanet.Net.Transports.ITransport.html
+[NetMQTransport]: https://docs.libplanet.io/0.19.0/api/Libplanet.Net.Transports.NetMQTransport.html
+[TcpTransport]: https://github.com/planetarium/libplanet/pull/1523
+[Swarm<T>]: https://docs.libplanet.io/0.19.0/api/Libplanet.Net.Swarm-1.html
+
+
+Configurable <abbr title="Proof-Of-Work">PoW</abbr> Hash Algorithm
+------------------------------------------------------------------
+
+*Added in version [0.12.0].*
+
+Previously, Libplanet has used [SHA-256] for proof-of-work.  However, since
+version [0.12.0], each network can configure its hash algorithm for
+<abbr title="proof-of-work">PoW</abbr> by implementing its
+[`IBlockPolicy<T>.GetHashAlgorithm()`
+method][IBlockPolicy<T>.GetHashAlgorithm].  As it is not only configurable
+for each network, but also for block indices, it can be migrated within
+a network after its establishment.
+
+You can use any hash algorithm for <abbr title="proof-of-work">PoW</abbr>
+as long as it is a subclass of the [`HashAlgorithm`][HashAlgorithm] abstract
+class, which is built in the .NET runtime.  It is not necessary to be included
+in the .NET runtime, but you can use your custom implementation as well.
+For instance, [RandomXSharp], a .NET binding library for RandomX we made,
+also provides a subclass of `HashAlgorithm`.
+
+The following code shows an example policy requiring blocks below 10,000th
+block to use SHA-256 and above blocks to use SHA-512 within a chain:
+
+~~~~ csharp
+HashAlgorithmType GetHashAlgorithm(long index) =>
+    index <= 10_000
+        ? HashAlgorithmType.Of<SHA256>()
+        : HashAlgorithmType.Of<SHA512>();
+~~~~
+
+[SHA-256]: https://ko.wikipedia.org/wiki/SHA-2
+[IBlockPolicy<T>.GetHashAlgorithm]: https://docs.libplanet.io/0.19.0/api/Libplanet.Blockchain.Policies.IBlockPolicy-1.html#Libplanet_Blockchain_Policies_IBlockPolicy_1_GetHashAlgorithm_System_Int64_
+[HashAlgorithm]: https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.hashalgorithm
+[RandomXSharp]: https://github.com/planetarium/RandomXSharp
+
+
+Non-blocking Renderers
+----------------------
+
+*Added in version [0.14.0].*
+
+[`IRenderer<T>`][IRenderer<T>] and [`IActionRenderer<T>`][IActionRenderer<T>]
+interfaces, event listners that wait for state transitions of node's local
+chain, flow in a blocking manner as they do not spawn new threads.  Suppose
```

Comment:
```suggestion
interfaces, event listeners that wait for state transitions of node's local
chain, do not spawn new threads but flow in a blocking manner.  Suppose
```
Placing the phrase `flow in a blocking manner` to the end of the sentence puts more emphasis on the blocking nature of the two interfaces which fits the intention of this section.